In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from scipy.io import loadmat

In [11]:

eeg = loadmat("../Dataset/dataset.mat")
rating = loadmat("../Dataset/rating.mat")
classes = loadmat("../Dataset/class_012.mat")

In [14]:
print(eeg.keys())
print(rating.keys())
print(classes.keys())

dict_keys(['__header__', '__version__', '__globals__', 'dataset'])
dict_keys(['__header__', '__version__', '__globals__', 'rating'])
dict_keys(['__header__', '__version__', '__globals__', 'class_012'])


| **Key**       | **Description**                                                                                    | **Purpose**                                                                |
| ------------- | -------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------- |
| `__header__`  | Contains metadata about the MATLAB file, such as the MATLAB version and file creation information. | Provides file information; not used in EEG analysis.                       |
| `__version__` | Stores the MATLAB file format version.                                                             | Indicates which MATLAB version was used to save the file.                  |
| `__globals__` | Contains global MATLAB variables (usually empty).                                                  | Generally not used for EEG analysis.                                       |
| `dataset`     | Contains the actual EEG recordings.                                                                | Main EEG data used for preprocessing, visualization, and machine learning. |
| `rating`      | Contains the participants' self-reported workload ratings (values from 4 to 9).                    | Used as the ground truth workload scores.                                  |
| `class_012`   | Contains the workload class labels for each participant (0 = Low, 1 = Medium, 2 = High).           | Used as target labels for classification models.                           |


In [16]:
print(eeg['dataset'].shape)
print(rating['rating'].shape)
print(classes['class_012'].shape)

(14, 19200, 45)
(45, 1)
(45, 1)


| File            | Shape               | Meaning                                            |
| --------------- | ------------------- | -------------------------------------------------- |
| `dataset.mat`   | **(14, 19200, 45)** | 14 EEG channels × 19,200 samples × 45 participants |
| `rating.mat`    | **(45, 1)**         | Self-reported workload rating for each participant |
| `class_012.mat` | **(45, 1)**         | Workload class label for each participant          |


In [17]:
print(rating['rating'])

[[8]
 [5]
 [5]
 [5]
 [7]
 [5]
 [7]
 [7]
 [6]
 [5]
 [6]
 [7]
 [7]
 [6]
 [9]
 [9]
 [7]
 [5]
 [8]
 [6]
 [8]
 [5]
 [4]
 [8]
 [4]
 [7]
 [6]
 [8]
 [6]
 [7]
 [5]
 [9]
 [4]
 [8]
 [8]
 [5]
 [8]
 [5]
 [6]
 [7]
 [8]
 [8]
 [7]
 [8]
 [7]]


In [18]:
print(classes['class_012'])

[[2]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [1]
 [1]
 [0]
 [1]
 [1]
 [1]
 [1]
 [2]
 [2]
 [1]
 [0]
 [2]
 [1]
 [2]
 [0]
 [0]
 [2]
 [0]
 [1]
 [1]
 [2]
 [1]
 [1]
 [0]
 [2]
 [0]
 [2]
 [2]
 [0]
 [2]
 [0]
 [1]
 [1]
 [2]
 [2]
 [1]
 [2]
 [1]]


In [19]:
print(eeg['dataset'])

[[[4584.62 4663.59 4212.82 ... 4233.33 4502.05 4105.13]
  [4584.1  4663.59 4213.85 ... 4225.13 4502.05 4109.23]
  [4574.36 4658.46 4228.21 ... 4213.85 4496.41 4115.9 ]
  ...
  [4605.64 4717.95 4168.72 ... 4216.41 4523.59 4227.18]
  [4606.15 4736.92 4186.67 ... 4220.51 4518.46 4223.59]
  [4611.28 4746.15 4202.05 ... 4231.79 4517.95 4215.38]]

 [[3902.05 3978.46 4148.72 ... 4220.   4781.03 4028.72]
  [3895.9  3965.64 4151.28 ... 4206.67 4778.46 4027.69]
  [3893.85 3965.64 4155.38 ... 4194.87 4769.23 4029.23]
  ...
  [3987.69 4144.62 4217.44 ... 4201.54 4770.77 4101.54]
  [3992.82 4164.1  4224.62 ... 4202.56 4773.33 4098.46]
  [4001.54 4134.36 4231.79 ... 4206.15 4779.49 4086.15]]

 [[4571.79 4641.03 3766.15 ... 3715.9  4411.79 4111.79]
  [4574.87 4643.08 3768.21 ... 3715.9  4412.31 4121.54]
  [4576.92 4641.54 3776.92 ... 3710.77 4407.18 4119.49]
  ...
  [4573.85 4631.28 3753.33 ... 3795.38 4452.82 4145.13]
  [4572.82 4636.92 3770.26 ... 3796.41 4447.18 4135.9 ]
  [4579.49 4631.79 3781.03

In [20]:
subject1 = eeg['dataset'][:, :, 0]

In [24]:
print(subject1.shape)

(14, 19200)


This means:

- 14 EEG channels
- 19,200 samples
- for Participant 1

In [26]:
print(eeg['dataset'].shape)
print(rating['rating'].shape)
print(classes['class_012'].shape)

print(rating['rating'])
print(classes['class_012'])

(14, 19200, 45)
(45, 1)
(45, 1)
[[8]
 [5]
 [5]
 [5]
 [7]
 [5]
 [7]
 [7]
 [6]
 [5]
 [6]
 [7]
 [7]
 [6]
 [9]
 [9]
 [7]
 [5]
 [8]
 [6]
 [8]
 [5]
 [4]
 [8]
 [4]
 [7]
 [6]
 [8]
 [6]
 [7]
 [5]
 [9]
 [4]
 [8]
 [8]
 [5]
 [8]
 [5]
 [6]
 [7]
 [8]
 [8]
 [7]
 [8]
 [7]]
[[2]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [1]
 [1]
 [0]
 [1]
 [1]
 [1]
 [1]
 [2]
 [2]
 [1]
 [0]
 [2]
 [1]
 [2]
 [0]
 [0]
 [2]
 [0]
 [1]
 [1]
 [2]
 [1]
 [1]
 [0]
 [2]
 [0]
 [2]
 [2]
 [0]
 [2]
 [0]
 [1]
 [1]
 [2]
 [2]
 [1]
 [2]
 [1]]


In [28]:
print(rating['rating'].T)
print(classes['class_012'].T)

[[8 5 5 5 7 5 7 7 6 5 6 7 7 6 9 9 7 5 8 6 8 5 4 8 4 7 6 8 6 7 5 9 4 8 8 5
  8 5 6 7 8 8 7 8 7]]
[[2 0 0 0 1 0 1 1 1 0 1 1 1 1 2 2 1 0 2 1 2 0 0 2 0 1 1 2 1 1 0 2 0 2 2 0
  2 0 1 1 2 2 1 2 1]]


| File              | Shape           | Description                  |
| ----------------- | --------------- | ---------------------------- |
| **dataset.mat**   | (14, 19200, 45) | EEG signals                  |
| **rating.mat**    | (45, 1)         | Workload ratings (4–9)       |
| **class_012.mat** | (45, 1)         | Three-class labels (0, 1, 2) |


In [29]:
X = eeg["dataset"]

y = rating["rating"]

In [30]:
print(type(X))
print(type(y))

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


Why is EEG stored as a NumPy array? 
- EEG is stored as a NumPy array because it is numerical, multi-dimensional data, and NumPy is the standard Python library for efficiently storing and processing large numerical datasets.

In [31]:
print(X.shape)

print(y.shape)

(14, 19200, 45)
(45, 1)


In [32]:
sampling_rate = 128

duration = X.shape[1] / sampling_rate

print(duration)

150.0


In [33]:
print(y)

[[8]
 [5]
 [5]
 [5]
 [7]
 [5]
 [7]
 [7]
 [6]
 [5]
 [6]
 [7]
 [7]
 [6]
 [9]
 [9]
 [7]
 [5]
 [8]
 [6]
 [8]
 [5]
 [4]
 [8]
 [4]
 [7]
 [6]
 [8]
 [6]
 [7]
 [5]
 [9]
 [4]
 [8]
 [8]
 [5]
 [8]
 [5]
 [6]
 [7]
 [8]
 [8]
 [7]
 [8]
 [7]]


In [34]:
np.unique(y)

array([4, 5, 6, 7, 8, 9], dtype=uint8)

In [35]:
binary_labels = np.where(y <= 6, 0, 1)

In [36]:
print(binary_labels)

[[1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [0]
 [0]
 [1]
 [1]
 [0]
 [1]
 [1]
 [1]
 [0]
 [1]
 [0]
 [1]
 [0]
 [0]
 [1]
 [0]
 [1]
 [0]
 [1]
 [0]
 [1]
 [0]
 [1]
 [0]
 [1]
 [1]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [1]
 [1]
 [1]
 [1]]


In [37]:
unique, counts = np.unique(binary_labels, return_counts=True)

print(dict(zip(unique, counts)))

{np.int64(0): np.int64(20), np.int64(1): np.int64(25)}


- 20 Low Workload subjects
- 25 High Workload subjects

In [38]:
np.save("binary_labels.npy", binary_labels)

| Question                                 | Answer                                                                                                                                |
| ---------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------- |
| Why is the subject dimension last?       | It is the organization chosen by the dataset creators in MATLAB, allowing each subject's EEG recording to be accessed as a 2D matrix. |
| Could the shape be 45 × 14 × 19200?      | Yes. The data can be reordered, and deep learning frameworks often prefer this format.                                                |
| Why 128 Hz?                              | The Emotiv EPOC headset samples at 128 times per second, which is sufficient for capturing important EEG frequency bands.             |
| Why 150 seconds?                         | Because 19,200 samples recorded at 128 Hz correspond to 150 seconds (2.5 minutes).                                                    |
| Why are ratings subjective?              | They are based on participants' perceived mental workload after the task.                                                             |
| Could two people have different ratings? | Yes. Individual differences such as experience, stress, and cognitive ability influence perceived workload.                           |
| Why choose 6 as the threshold?           | It follows the labeling strategy used in previous STEW studies, enabling fair comparison with published work.                         |
| Could another threshold work?            | Yes, but it would change the class definitions and reduce comparability with prior research.                                          |
| Are both classes equally represented?    | They are not perfectly equal but are reasonably balanced (20 low vs. 25 high).                                                        |
| Should class balancing be applied?       | Not initially. The slight imbalance is unlikely to bias the models significantly.                                                     |


## Module 2: EEG Signal Visualization